File: examples/mobile/ios/Sources/PhantomDemoKit/PhantomServerConfig.swift
// SPDX-License-Identifier: Apache-2.0
//
// PhantomServerConfig — the server endpoint plus its pinned hybrid verifying
// key. Server identity pinning is mandatory (Security Invariant 1): the pinned
// key is loaded from a bundled resource baked into the app at build time and is
// NEVER fetched at runtime. Fetching the key over the network would void the
// trust model entirely.

import Foundation

/// Failure modes when resolving the server config.
public enum ConfigError: Error, CustomStringConvertible {
    case pinnedKeyResourceMissing
    case pinnedKeyEmpty
    case malformedHex

    public var description: String {
        switch self {
        case .pinnedKeyResourceMissing:
            return "phantom_server_pk.bin not found in the app bundle"
        case .pinnedKeyEmpty:
            return "phantom_server_pk.bin was empty"
        case .malformedHex:
            return "pinned key hex string is malformed"
        }
    }
}

/// Immutable description of the Phantom Protocol server to connect to.
public struct PhantomServerConfig {
    public let host: String
    public let port: UInt16
    /// The server's hybrid verifying key (Ed25519 || ML-DSA-65 public halves),
    /// exactly the bytes produced by `PhantomListener::verifying_key_bytes()` /
    /// the admin CLI `pubkey` command.
    public let pinnedKey: Data

    public init(host: String, port: UInt16, pinnedKey: Data) {
        self.host = host
        self.port = port
        self.pinnedKey = pinnedKey
    }

    // MARK: - Bundled-resource loader

    /// Name of the bundled resource holding the raw pinned key bytes.
    public static let pinnedKeyResource = "phantom_server_pk"
    public static let pinnedKeyExtension = "bin"

    /// Loads the pinned key from the app bundle. Pass the bundle that actually
    /// carries the resource — in this sample it is bundled with `PhantomDemoKit`,
    /// so `Bundle.module` is the right default.
    ///
    /// If the bundled resource is missing or empty, falls back to the
    /// `developmentPinnedKeyHex` constant below so the sample is runnable out of
    /// the box against a locally-keyed dev server. In a shipping app you should
    /// treat a missing bundled key as a fatal misconfiguration instead.
    public static func loadPinnedKey(from bundle: Bundle = .module) throws -> Data {
        if let url = bundle.url(forResource: pinnedKeyResource, withExtension: pinnedKeyExtension),
           let data = try? Data(contentsOf: url),
           !data.isEmpty,
           !looksLikePlaceholder(data) {
            return data
        }

        // Fallback for the demo only. Replace developmentPinnedKeyHex with the
        // hex emitted by `cargo run --manifest-path cli/Cargo.toml -- pubkey ...`
        // (or bake real bytes into phantom_server_pk.bin and this branch is never
        // taken). See README.md.
        let fallback = try parseHex(developmentPinnedKeyHex)
        guard !fallback.isEmpty else { throw ConfigError.pinnedKeyEmpty }
        return fallback
    }

    /// Convenience: build a config for the given endpoint using the bundled
    /// pinned key.
    public static func bundled(host: String,
                               port: UInt16,
                               bundle: Bundle = .module) throws -> PhantomServerConfig {
        let key = try loadPinnedKey(from: bundle)
        return PhantomServerConfig(host: host, port: port, pinnedKey: key)
    }

    // MARK: - Dev fallback key (REPLACE ME)
    //
    // This is a clearly-marked placeholder. The all-zero hex is recognised by
    // `looksLikePlaceholder` and treated as "no key bundled". Paste the real
    // pubkey hex here for a zero-config dev run, OR bake phantom_server_pk.bin.
    //
    // NEVER fetch this value at runtime — pinning is the whole security model.
    public static let developmentPinnedKeyHex =
        "0000000000000000000000000000000000000000000000000000000000000000"

    /// A bundled key consisting entirely of 0x00 is the documented placeholder;
    /// treat it as "not configured".
    static func looksLikePlaceholder(_ data: Data) -> Bool {
        !data.isEmpty && data.allSatisfy { $0 == 0 }
    }

    // MARK: - Hex helper

    /// Parses an even-length hex string (no `0x`, whitespace tolerated) into
    /// bytes. Used to turn the admin CLI's `keygen`/`pubkey` hex output into the
    /// pinned-key `Data` for the README workflow.
    public static func parseHex(_ hex: String) throws -> Data {
        let cleaned = hex.filter { !$0.isWhitespace }
        guard cleaned.count % 2 == 0 else { throw ConfigError.malformedHex }
        var bytes = Data()
        bytes.reserveCapacity(cleaned.count / 2)
        var index = cleaned.startIndex
        while index < cleaned.endIndex {
            let next = cleaned.index(index, offsetBy: 2)
            guard let byte = UInt8(cleaned[index..<next], radix: 16) else {
                throw ConfigError.malformedHex
            }
            bytes.append(byte)
            index = next
        }
        return bytes
    }
}

File: examples/mobile/android/app/src/main/kotlin/com/example/phantomdemo/PhantomServerConfig.kt
// SPDX-License-Identifier: Apache-2.0
package com.example.phantomdemo

import android.content.Context

/**
 * Connection parameters for the demo server.
 *
 * The pinned verifying key is the heart of the trust model (Security Invariant
 * 1): the client refuses to establish a session unless the server proves
 * possession of the private half of [pinnedKey]. The key is the public output
 * of `PhantomListener::verifying_key_bytes()` on the server, surfaced through
 * the CLI `pubkey` command.
 *
 * The key MUST be baked into the app at build time — never fetched over the
 * network at runtime, which would void the pinning guarantee. Rotating the
 * server signing key therefore requires shipping an app update.
 */
data class PhantomServerConfig(
    val host: String,
    val port: UShort,
    val pinnedKey: ByteArray,
) {
    init {
        require(pinnedKey.isNotEmpty()) { "pinned key must not be empty" }
    }

    // data class with a ByteArray field needs hand-written equals/hashCode.
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PhantomServerConfig) return false
        return host == other.host &&
            port == other.port &&
            pinnedKey.contentEquals(other.pinnedKey)
    }

    override fun hashCode(): Int {
        var result = host.hashCode()
        result = 31 * result + port.hashCode()
        result = 31 * result + pinnedKey.contentHashCode()
        return result
    }

    companion object {
        /** The demo server host. Override for your own deployment. */
        const val DEMO_HOST: String = "10.0.2.2" // emulator host loopback

        /** Matches `phantom-server`'s default listen port. */
        val DEMO_PORT: UShort = 4242.toUShort()

        /**
         * Demo-only placeholder pinned key, expressed as hex. This is a clean,
         * obvious 32-byte all-zero placeholder (64 hex chars) — it is NOT a real
         * verifying key and is NOT even the right length: a real
         * `HybridVerifyingKey` (Ed25519 || ML-DSA-65 public halves) is 1984
         * bytes. It exists only so the app compiles and runs without first
         * dropping a `res/raw/phantom_server_pk` file. A connect attempt with
         * this value will (correctly) fail to parse/pin against any real server.
         * Replace `res/raw/phantom_server_pk` with the real 1984 bytes from the
         * CLI `pubkey` command.
         */
        private const val DEMO_PINNED_KEY_HEX: String =
            "0000000000000000000000000000000000000000000000000000000000000000"

        /**
         * Build the config. Prefers the baked-in `res/raw/phantom_server_pk`
         * (raw verifying-key bytes); falls back to [DEMO_PINNED_KEY_HEX] when
         * that resource is absent so the project still builds out of the box.
         */
        fun load(
            context: Context,
            host: String = DEMO_HOST,
            port: UShort = DEMO_PORT,
        ): PhantomServerConfig {
            val keyBytes = loadPinnedKey(context)
            return PhantomServerConfig(host = host, port = port, pinnedKey = keyBytes)
        }

        /**
         * Read the pinned verifying key from `res/raw/phantom_server_pk`. The
         * resource is the raw byte output of the CLI `pubkey` command (NOT hex).
         * Falls back to the documented demo hex constant when the resource is
         * missing or empty.
         */
        private fun loadPinnedKey(context: Context): ByteArray {
            val resId = context.resources.getIdentifier(
                "phantom_server_pk",
                "raw",
                context.packageName,
            )
            if (resId != 0) {
                val bytes = context.resources.openRawResource(resId).use { it.readBytes() }
                if (bytes.isNotEmpty()) {
                    return bytes
                }
            }
            return hexToBytes(DEMO_PINNED_KEY_HEX)
        }

        /** Decode an even-length hex string into bytes. */
        fun hexToBytes(hex: String): ByteArray {
            val clean = hex.trim().removePrefix("0x")
            require(clean.length % 2 == 0) { "hex string must have an even length" }
            val out = ByteArray(clean.length / 2)
            for (i in out.indices) {
                val hi = Character.digit(clean[i * 2], 16)
                val lo = Character.digit(clean[i * 2 + 1], 16)
                require(hi >= 0 && lo >= 0) { "invalid hex character at index ${i * 2}" }
                out[i] = ((hi shl 4) or lo).toByte()
            }
            return out
        }
    }
}
